[JENKINS-27624][JENKINS-42709] Bump Jenkins baseline to Java 8 minimum

batmat opened the pull request:
# Description

Cf. article on jenkins.io: [Jenkins Upgrades To Java 8](https://jenkins.io/blog/2017/01/17/Jenkins-is-upgrading-to-Java-8/)

See also [JENKINS-42709](https://issues.jenkins-ci.org/browse/JENKINS-42709).

Details: 

* [X] Fix generics issues
* [X] Fix infradna/bridge-method-injector to work with Jenkins@JDK8 
* [x] Fix compilation error (in test module)

```
[INFO] 100 warnings
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/tiste/dev/JENKINS/jenkins/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```
I guess some help from @kohsuke may be useful here. cc @jglick & @stephenc too.

<!-- Comment: 
If the issue is not fully described in the ticket, please put additional comments (justification, pull request links, etc.).
-->

### Changelog entries

Proposed changelog entries:

* JENKINS-42709: Make Jenkins buildable with JDK 8
* ...

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [X] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

<!-- Comment: 
 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Desired reviewers

@mention

<!-- Comment:
If you want to get reviews from particular people, please CC them.
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

szpak committed:
[JENKINS-42709] Unable to build for Java 8

With Java 8 set as build target version.

batmat committed:
Make it compilable with JDK8

batmat committed:
Clarify PermissionGroup.owner is @Nonnull

Permission.owner is already marked @Nonnull, and
is built using PermissionGroup.owner.

jglick commented:
```java
        // checks the class version
        if (readShort(off + 6) > Opcodes.V1_7) {
            throw new IllegalArgumentException();
        }
```

Did you try just upgrading `asm-debug-all` from 4.2, say to 5.2?

```java
        // checks the class version
        if (readShort(off + 6) > Opcodes.V1_8) {
            throw new IllegalArgumentException();
        }
```

batmat commented:
@jglick Pretty sure I did try that, among other combinations. But I will retry to make sure.

batmat commented:
To sum up, the hint of Jesse above pushed me into the right direction. I think I had touched Jenkins' pom.xml instead of the plugin's when twiddling with it.

So, now I'm past it, the good news is that the WAR build succeeds (though I didn't smoke test it yet), only the `test` module fails because of a Javac error. I think this is coming from the inference fixes that have been added while introducing lambdas, and now the code is ambiguous. Looking into it.

https://github.com/jenkinsci/jenkins/blob/749346896abbdce8e074d593ab37175f59a84c29/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java#L89 does not compile, and shows:
```
[ERROR] /home/tiste/dev/JENKINS/jenkins/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
[INFO] 1 error
```

batmat committed:
Fix bridge-method-injector IllegalArgumentException

```
[ERROR] Failed to execute goal com.infradna.tool:bridge-method-injector:1.13:process (default) on project jenkins-core: Failed to process @WithBridgeMethods: Failed to process /home/tiste/dev/tmp/2017-03-29T23h05m22+0200-jenkins-core/jenkins/jenkins/core/target/classes/hudson/model/AbstractItem.class: IllegalArgumentException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.infradna.tool:bridge-method-injector:1.13:process (default) on project jenkins-core: Failed to process @WithBridgeMethods
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to process @WithBridgeMethods
        at com.infradna.tool.bridge_method_injector.ProcessMojo.execute(ProcessMojo.java:68)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
        ... 20 more
Caused by: java.io.IOException: Failed to process /home/tiste/dev/tmp/2017-03-29T23h05m22+0200-jenkins-core/jenkins/jenkins/core/target/classes/hudson/model/AbstractItem.class
        at com.infradna.tool.bridge_method_injector.MethodInjector.handle(MethodInjector.java:106)
        at com.infradna.tool.bridge_method_injector.ProcessMojo.execute(ProcessMojo.java:65)
        ... 22 more
Caused by: java.lang.IllegalArgumentException
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:170)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:153)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:424)
        at com.infradna.tool.bridge_method_injector.MethodInjector.handle(MethodInjector.java:74)
        ... 23 more
[ERROR]

```

batmat committed:
Fix generics ambiguity

Inline the method to use concrete subtypes of AbstractBuild to remove ambiguity

```
[INFO] Jenkins war ........................................ SUCCESS [04:19 min]
[INFO] Tests for Jenkins core ............................. FAILURE [ 18.740 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 06:01 min
[INFO] Finished at: 2017-03-29T23:07:16+02:00
[INFO] Final Memory: 92M/504M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.0:testCompile (default-testCompile) on project test: Compilation failure
[ERROR] /home/tiste/dev/JENKINS/jenkins/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :test
```

batmat commented:
So, now it fails with the following only in CI. And not locally with `1.8.0_121`.
I can reproduce using `1.8.0_66`, so I suspect it's a javac bug that got fixed between the two.

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
[INFO] 1 error
```

batmat committed:
Fix error with 1.8.0_77 in CI, does not happen with 1.8.0_121

Probably a bit ugly, but well for a test it should be acceptable.

batmat commented:
@jenkinsci/code-reviewers review welcome: this PR bumps Jenkins core to JDK8 minimum.

aheritier commented:
Maybe https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core should be removed ?

jglick commented on pom.xml:
```diff
@@ -483,6 +483,15 @@ THE SOFTWARE.
           <groupId>com.infradna.tool</groupId>
           <artifactId>bridge-method-injector</artifactId>
           <version>1.13</version>
+          <!-- workaround waiting for a release with https://github.com/infradna/bridge-method-injector/pull/14 -->
```
1.15 has been released acc. to @kohsuke.

batmat replied:
Fixed with https://github.com/jenkinsci/jenkins/pull/2802/commits/496d574b8c31f06b3c005cbdc84370bb0d47c69a

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -655,7 +655,7 @@ private boolean relatedTo(@Nonnull AbstractBuild<?,?> b) {
     @WithBridgeMethods(List.class)
     public @Nonnull RunList getBuilds() {
-        return RunList.fromJobs(Jenkins.getInstance().allItems(Job.class)).filter(new Predicate<Run<?,?>>() {
+        return RunList.fromJobs((Iterable)Jenkins.getInstance().allItems(Job.class)).filter(new Predicate<Run<?,?>>() {
```
Might suffice to just replace `Run<?,?>` with `Run` (untested). Or to add a qualifier to `filter` like

```java
return RunList.fromJobs(Jenkins.getInstance().allItems(Job.class)).<Something<Here>>filter(new Predicate<Run<?,?>>() {
```

batmat replied:
I had tried quite a few combinations before finding that one, for instance with `Run` only:

![image](https://cloud.githubusercontent.com/assets/223853/24515358/a9dcacf2-1577-11e7-95fb-6eae5f7d1aa1.png)

So I'm ready to change to something else, but I thought the fix I have was readable/simple enough anyway (?).

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -778,7 +778,7 @@ public String getUrl() {
 
     public RunList getBuilds() {
-        return RunList.fromJobs(Jenkins.getInstance().allItems(Job.class)).node(getNode());
+        return RunList.fromJobs((Iterable)Jenkins.getInstance().allItems(Job.class)).node(getNode());
```
Ditto.

oleg-nenashev commented on test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:
```diff
@@ -17,7 +17,7 @@
 
     @Test
-    public void isJenkinsJVM() throws Exception {
+    public void isJenkinsJVM() throws Throwable {
```
Hm? Why would you need such change?

batmat replied:
@oleg-nenashev I tried to separate each "fix" in a dedicated commit. This is https://github.com/jenkinsci/jenkins/pull/2802/commits/77dfa64b6a173b58c7f61107dcc51a61f7a3466a

batmat replied:
The build logs exhibiting the issue: https://ci.jenkins.io/job/Core/job/jenkins/job/PR-2802/2/execution/node/28/log/

Copying the end for future ref when the build has been wiped out:

```
proprietary API and may be removed in a future release
21:18:06 [INFO] 100 warnings 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [ERROR] COMPILATION ERROR : 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
21:18:06 [INFO] 2 errors 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] Reactor Summary:
21:18:06 [INFO] 
21:18:06 [INFO] Jenkins main module ................................ SUCCESS [ 45.970 s]
21:18:06 [INFO] Jenkins cli ........................................ SUCCESS [ 42.146 s]
21:18:06 [INFO] Jenkins core ....................................... SUCCESS [03:12 min]
21:18:06 [INFO] Jenkins war ........................................ SUCCESS [01:23 min]
21:18:06 [INFO] Tests for Jenkins core ............................. FAILURE [ 56.812 s]
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] BUILD FAILURE
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] Total time: 07:18 min
21:18:06 [INFO] Finished at: 2017-03-29T21:18:05+00:00
21:18:06 [INFO] Final Memory: 100M/594M
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.0:testCompile (default-testCompile) on project test: Compilation failure: Compilation failure:
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
21:18:06 [ERROR] -> [Help 1]
21:18:06 [ERROR] 
21:18:06 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
21:18:06 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
21:18:06 [ERROR] 
21:18:06 [ERROR] For more information about the errors and possible solutions, please read the following articles:
21:18:06 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
21:18:06 [ERROR] 
21:18:06 [ERROR] After correcting the problems, you can resume the build with the command
21:18:06 [ERROR]   mvn <goals> -rf :test
```

batmat replied:
@oleg-nenashev does this address your question/concern? Thanks

oleg-nenashev replied:
it does

batmat committed:
Use released bridge-method-injector:1.15

And suppress associated workaround

batmat commented:
@olivergondza I would really appreciate your review, especially with your LTS Officer cap, as this will possibly propagate to next one (or next-next) if we merge it soonish as initially planned.

oleg-nenashev reviewed (approved):
I am fine with the PR itself, but maybe it makes sense to postpone it till the next weekly. #2826 is a serious regression, and I would rather allow Java 7 users to pick it. Not a blocker, just a quite biased IMHO

daniel-beck commented:
Could we get rid of occurrences of `MaxPermSize` then? It means nothing on JDK 8 but spams the logs.

batmat committed:
Remove MaxPermSize usage

Obsolete with Java 8. Permanent generation was removed.

batmat commented:
@daniel-beck done. Arnaud had already asked for it, and I said I would do it in a followup PR (to keep that PR a bit _shorter_). As you were now two to request it, I went ahead and did it here.

batmat commented:
I believe all concerns have been addressed and that this pull request is ready to merge.

If nobody objects in the meantime, or does it before me, I plan to merge this PR before the end of the week. 

**Which means that the next weekly will then require Java 8 to run.**

szpak reviewed (approved):
LGTM

jglick commented:
> #2826 is a serious regression, and I would rather allow Java 7 users to pick it.

@oleg-nenashev perhaps you missed our policy update when getting rid of the `rc` branch, but in summary: if you find a serious regression and fix it, you can ask @kohsuke to cut an out-of-order trunk release. No need to wait a week.

oleg-nenashev commented:
@jglick Asn a requester of several out-of-order releases on this year, I definitely know about this opportunity. And I also know about obstacles like KK's reachability. In the case of this fix it has landed too late (late Friday) for making out-of-order release meaningful. 

BTW I think we should just go forward with this PR

oleg-nenashev commented:
Merging without squash since there are 2 authors && the change is not a backporting candidate

oleg-nenashev commented:
Uh oh, the merge button dissapeared :D

batmat commented:
Thank you everyone! \o/
